etc/dbus-serialbattery/sinowealth.py

```diff
@@ -0,0 +1,266 @@
+        if pack_voltage_data is False:
+            return False
+        pack_voltage = unpack_from('>H', pack_voltage_data[:-1])
+        logger.info(">>> INFO: current pack voltage: %f", pack_voltage[0]/1000)
+        self.voltage = pack_voltage[0]/1000
+        return True
+
+    def read_pack_current(self):
+        current_data = self.read_serial_data_sinowealth(self.command_current)
+        if current_data is False:
+            return False
+        current = unpack_from('>i', current_data[:-1])
+        logger.info(">>> INFO: current pack current: %f", current[0]/1000)
+        self.current = current[0]/1000
+        return True
+        
+    def read_remaining_capacity(self):
+        remaining_capacity_data = self.read_serial_data_sinowealth(self.command_remaining_capacity)
+        if remaining_capacity_data is False:
+            return False
+        remaining_capacity = unpack_from('>i', remaining_capacity_data[:-1])
+        logger.info(">>> INFO: remaining battery capacity: %f Ah", remaining_capacity[0]/1000)
+        self.capacity_remain = remaining_capacity[0]/1000
+        if self.capacity is None:
+          self.read_capacity()
+        self.total_ah_drawn = self.capacity - self.capacity_remain
+        return True
+ 
+    def read_capacity(self):
+        capacity_data = self.read_serial_data_sinowealth(self.command_capacity)
+        if capacity_data is False:
+            return False
+        capacity = unpack_from('>i', capacity_data[:-1])
+        logger.info(">>> INFO: Battery capacity: %f Ah", capacity[0]/1000)
+        self.capacity = capacity[0]/1000
+        return True       
+        
+    def read_pack_config_data(self):
+        # TODO: detect correct chipset, currently the pack_config_map register is parsed as,
+        # SH367303 / 367305 / 367306 / 39F003 / 39F004 / BMS_10. So these are the currently supported chips
+        pack_config_data = self.read_serial_data_sinowealth(self.command_pack_config)
+        if pack_config_data is False:
+            return False
+        cell_cnt_mask = int(7)
+        self.cell_count = (pack_config_data[1] & cell_cnt_mask) + 3
+        logger.info(">>> INFO: Number of cells: %u", self.cell_count)
+        temp_sens_mask = int(~(1 << 6))
+        self.temp_sensors = 1 if (pack_config_data[1] & temp_sens_mask) else 2 # one means two
+        logger.info(">>> INFO: Number of temperatur sensors: %u", self.temp_sensors)
+        return True
+        
+    def read_cell_data(self):
+        if self.cell_count is None:
+          self.read_pack_config_data()
+          
+        cell_index = 1
+        while cell_index <= self.cell_count:
+          self.cell_voltages[cell_index] = self.read_cell_voltage(cell_index)
+          if self.cell_voltages[cell_index] is False:
+            return False
+          cell_index += 1
+        
+        self.cell_max_no = max(self.cell_voltages, key=int) - 1
```

Comment:
If you use the cells object that exist in battery.py instead of your own cell_voltages then the min/max values will be done for you in battery as well.